### PR TITLE
P2 233 be create farm

### PR DESCRIPTION
**Pull request by BryanCaldeira**
Internal usage routes for creation of Farm, Update to token to have Farm info, User Schema changes to have reference field Farm

### CODE REVIEW

File: controllers/authentication.controllers.ts [hunk 1]
```diff
@@ -134,19 +159,19 @@ async function signIn (req: Request, res: Response) {
       id: user._id,
       name: user.name,
       email: user.email,
-      farmId: user.farmId,
+      farm: user.farm,
     });
```
Comment (BryanCaldeira) [inline controllers/authentication.controllers.ts:163]: Token will have farm info instead of farmId

File: controllers/farm.controllers.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+const farmController = {
+  create,
+  getAll,
+  getById,
+  updateById,
+  deleteById,
+};
+
+export default farmController;
```
Comment (BryanCaldeira) [inline controllers/farm.controllers.ts:175]: CRUD operations for FARM

File: middleware/authorization.middleware.ts [hunk 3]
```diff
@@ -6,10 +6,19 @@ import {
...
+  "farms"
+];
+
 // all requests pass through this function to check "Authorization" token
 async function authUser (req: Request, res: Response, next: NextFunction) {
-  // auth paths, security not needed
-  if (req.url.includes("auth")) {
+  // check for ignore paths
+  const isIgnorePath = ignoreAuthRoutes.some(route => req.url.split("/")?.at(3)?.toLocaleLowerCase("en-US") === route);
+
```
Comment (BryanCaldeira) [inline middleware/authorization.middleware.ts:18]: If params had "?auth" the validation was true, This was a major flaw which could bypass security checks

Comment (BryanCaldeira) [reply middleware/authorization.middleware.ts:18]: This has been resolved by taking the exact initial URL route to match and ignore certain routes

File: models/Farm.ts [hunk 4]
```diff
@@ -7,8 +7,9 @@ const Farm = model<IFarmSchema>(
   new Schema(
     {
       name: { type: String, required: true },
-      address: { type: String, required: true },
-      userIds: { type: [String], required: true },
+      address: { type: String, default: null },
+      users: { type: [Schema.Types.ObjectId], ref: "User", default: [] },
+      isDisabled: { type: Boolean, default: false }
```
Comment (BryanCaldeira) [inline models/Farm.ts:12]: Farm models will have reference field of User

File: models/User.ts [hunk 5]
```diff
@@ -3,22 +3,28 @@ import { Schema, model } from "mongoose";
...
-);
+});
+
+// hide password when converting to JSON
+UserSchema.methods.toJSON = function () {
+  const userObject = this.toObject();
+  delete userObject.password;
+  return userObject;
+}
+
```
Comment (BryanCaldeira) [inline models/User.ts:27]: When User schema is converted to JSON we need to delete the `password` field for security

**Review by whamester — APPROVED** [review]

**Review by ReinhardtBotha — APPROVED** [review]
🦓

**Review by ALFAROSO — APPROVED** [review]

**Pull request merged**